// @lux-ai/2021-challenge/lib/es6/GameMap/cell.js
import { Resource } from '../Resource';
import { CityTile } from '../Game/city';
import { Position } from './position';
/**
 * Cell class for map cells
 *
 * Some restrictions not explicitly employed:
 * Cell can either be empty (no resource or citytile), or have a resource, or have a citytile, not multiple.
 * There may be multiple units but this is only allowed on city tiles
 */
export class Cell {
    constructor(x, y, configs) {
        this.configs = configs;
        this.resource = null;
        this.citytile = null;
        /** map from unit id to the unit on this tile */
        this.units = new Map();
        /** How much a units cooldown goes down on this tile. This is higher if there are more developed roads */
        this.road = 0;
        this.pos = new Position(x, y);
        this.road = this.configs.parameters.MIN_ROAD;
    }
    /**
     * Set resource at cell and the amount of it
     * @param resourceType
     * @param amount
     */
    setResource(resourceType, amount) {
        this.resource = new Resource(resourceType, amount);
        return this.resource;
    }
    hasResource() {
        return this.resource !== null && this.resource.amount > 0;
    }
    /** Marks this as a city tile with the specified team */
    setCityTile(team, cityid) {
        this.citytile = new CityTile(team, this.configs);
        this.citytile.pos = this.pos;
        this.citytile.cityid = cityid;
    }
    isCityTile() {
        return this.citytile !== null;
    }
    hasUnits() {
        return this.units.size !== 0;
    }
    getRoad() {
        if (this.isCityTile()) {
            return this.configs.parameters.MAX_ROAD;
        }
        else {
            return this.road;
        }
    }
}
