// dimensions-ai/lib/main/SupportedPlugins/FileSystemStorage/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Storage_1 = require("../../Plugin/Storage");
var Dimension_1 = require("../../Dimension");
var Plugin_1 = require("../../Plugin");
var path_1 = __importDefault(require("path"));
var System_1 = require("../../utils/System");
var LRUFileCache_1 = __importDefault(require("../../utils/LRUFileCache"));
var Logger_1 = require("../../Logger");
var DeepMerge_1 = require("../../utils/DeepMerge");
var DeepCopy_1 = require("../../utils/DeepCopy");
var fs_1 = require("fs");
var FileSystemStorage = /** @class */ (function (_super) {
    __extends(FileSystemStorage, _super);
    function FileSystemStorage(configs) {
        var _this = _super.call(this) || this;
        _this.name = 'FS-Storage';
        _this.type = Plugin_1.Plugin.Type.STORAGE;
        _this._useCacheCount = 0;
        _this.configs = {
            maxCacheSize: 1024 * 1024 * 50,
            cacheDir: 'cache',
            loggingLevel: Logger_1.Logger.LEVEL.INFO,
        };
        _this.configs = DeepMerge_1.deepMerge(_this.configs, DeepCopy_1.deepCopy(configs));
        _this.lruFileCache = new LRUFileCache_1.default(_this.configs.maxCacheSize, path_1.default.join(System_1.LOCAL_DIR, _this.configs.cacheDir));
        _this.log = new Logger_1.Logger(_this.configs.loggingLevel, 'FS-Storage');
        return _this;
    }
    /**
     * Initializer. Initializes the storage object and creates necessary buckets
     */
    FileSystemStorage.prototype.initialize = function (dimension) {
        return __awaiter(this, void 0, void 0, function () {
            var bucketPath;
            return __generator(this, function (_a) {
                bucketPath = path_1.default.join(System_1.LOCAL_DIR, dimension.name.toLowerCase().replace(/ /g, '_') +
                    '_' +
                    dimension.id.toLowerCase());
                this.bucketPath = bucketPath;
                return [2 /*return*/];
            });
        });
    };
    FileSystemStorage.prototype.writeFileToBucket = function (file, dest) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.log.system("writing to bucket " + file + " -> " + dest);
                        return [4 /*yield*/, System_1.writeFileToDestination(file, path_1.default.join(this.bucketPath, dest))];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    FileSystemStorage.prototype.writeFileFromBucket = function (key, dest) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.log.system("writing from bucket " + key + " -> " + dest);
                        return [4 /*yield*/, System_1.writeFileToDestination(path_1.default.join(this.bucketPath, key), dest)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    FileSystemStorage.prototype.uploadTournamentFile = function (file, user, tournament) {
        return __awaiter(this, void 0, void 0, function () {
            var dest;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        dest = "users/" + user.username + "_" + user.playerID + "/tournaments/" + tournament.getKeyName() + "/bot.zip";
                        return [4 /*yield*/, this.writeFileToBucket(file, dest)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, dest];
                }
            });
        });
    };
    FileSystemStorage.prototype.upload = function (file, destination) {
        return __awaiter(this, void 0, void 0, function () {
            var dest;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        dest = "" + (destination ? destination : path_1.default.basename(file));
                        return [4 /*yield*/, this.writeFileToBucket(file, dest)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, dest];
                }
            });
        });
    };
    FileSystemStorage.prototype.download = function (key, destination, useCached) {
        return __awaiter(this, void 0, void 0, function () {
            var cachedPath_1, cachedPath;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (useCached) {
                            cachedPath_1 = this.lruFileCache.get(key);
                            // if there is a cached path, use it
                            if (cachedPath_1) {
                                this._useCacheCount++;
                                // TODO: It would be nice if we could just resolve with the cachedPath and let user use that, but
                                // unfortunately that invokes a race condition where downloading two files causes one file's cache to be
                                // invalidated and thus unretrievable after we give the user the old cachedPath. This race condition shouldn't
                                // happen often, if ever unless there are very little bots and max size is low.
                                // a better optimization would be to
                                fs_1.copyFileSync(cachedPath_1, destination);
                                return [2 /*return*/, destination];
                            }
                        }
                        return [4 /*yield*/, this.writeFileFromBucket(key, destination)];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.lruFileCache.add(key, destination)];
                    case 2:
                        cachedPath = _a.sent();
                        this.log.system(key + " cached to " + cachedPath);
                        return [2 /*return*/, destination];
                }
            });
        });
    };
    /**
     * Returns a download URL to use to download an object
     * @param key - key referencing the object to download
     */
    FileSystemStorage.prototype.getDownloadURL = function (key) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, path_1.default.join(this.bucketPath, key)];
            });
        });
    };
    FileSystemStorage.prototype.manipulate = function (dimension) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                dimension.configs.backingStorage = Dimension_1.StorageType.FS;
                return [2 /*return*/];
            });
        });
    };
    return FileSystemStorage;
}(Storage_1.Storage));
exports.FileSystemStorage = FileSystemStorage;
