// @lux-ai/2021-challenge/lib/es5/Actionable/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Actionable = void 0;
// any entity that has actions that are "a per unit" basis (not global) must extends this
var Actionable = /** @class */ (function () {
    function Actionable(configs) {
        this.configs = configs;
        // array of actions to handle in turn function
        this.currentActions = [];
        /** cooldown for this object before it can act */
        this.cooldown = 0;
    }
    /**
     * Checks if the cooldown is low enough to act again
     */
    Actionable.prototype.canAct = function () {
        return this.cooldown < 1;
    };
    /**
     * Game runs this function
     */
    Actionable.prototype.handleTurn = function (game) {
        try {
            this.turn(game);
        }
        finally {
            this.currentActions = [];
        }
        // reset actions to empty
    };
    Actionable.prototype.giveAction = function (action) {
        this.currentActions.push(action);
    };
    return Actionable;
}());
exports.Actionable = Actionable;
