// @lux-ai/2021-challenge/lib/es6/Actions/index.js
import { Unit } from '../Unit';
/**
 * Internal representations of agent sent actions. All actions must be validated prior to construction
 */
export class Action {
    constructor(action, team) {
        this.action = action;
        this.team = team;
    }
}
export class MoveAction extends Action {
    constructor(action, team, unitid, direction, newcell) {
        super(action, team);
        this.unitid = unitid;
        this.direction = direction;
        this.newcell = newcell;
    }
}
export class SpawnAction extends Action {
    constructor(action, team, x, y) {
        super(action, team);
        this.x = x;
        this.y = y;
    }
}
export class SpawnCartAction extends SpawnAction {
    constructor(action, team, x, y) {
        super(action, team, x, y);
        this.type = Unit.Type.CART;
    }
}
export class SpawnWorkerAction extends SpawnAction {
    constructor(action, team, x, y) {
        super(action, team, x, y);
        this.type = Unit.Type.WORKER;
    }
}
export class SpawnCityAction extends Action {
    constructor(action, team, unitid) {
        super(action, team);
        this.unitid = unitid;
    }
}
export class TransferAction extends Action {
    constructor(action, team, srcID, destID, resourceType, amount) {
        super(action, team);
        this.srcID = srcID;
        this.destID = destID;
        this.resourceType = resourceType;
        this.amount = amount;
    }
}
export class PillageAction extends Action {
    constructor(action, team, unitid) {
        super(action, team);
        this.unitid = unitid;
    }
}
export class ResearchAction extends Action {
    constructor(action, team, x, y) {
        super(action, team);
        this.x = x;
        this.y = y;
    }
}
