// @lux-ai/2021-challenge/lib/es5/GameMap/position.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Position = void 0;
var Game_1 = require("../Game");
var Position = /** @class */ (function () {
    function Position(x, y) {
        this.x = x;
        this.y = y;
    }
    Position.fromString = function (string) {
        return Position.fromArray(string.split(",").map(function (str) { return parseInt(str); }));
    };
    Position.fromArray = function (arr) {
        return new Position(arr[0], arr[1]);
    };
    Position.prototype.isAdjacent = function (pos) {
        var dx = this.x - pos.x;
        var dy = this.y - pos.y;
        if (Math.abs(dx) + Math.abs(dy) > 1) {
            return false;
        }
        return true;
    };
    Position.prototype.translate = function (direction, units) {
        if (units === void 0) { units = 1; }
        switch (direction) {
            case Game_1.Game.DIRECTIONS.NORTH:
                return new Position(this.x, this.y - units);
            case Game_1.Game.DIRECTIONS.EAST:
                return new Position(this.x + units, this.y);
            case Game_1.Game.DIRECTIONS.SOUTH:
                return new Position(this.x, this.y + units);
            case Game_1.Game.DIRECTIONS.WEST:
                return new Position(this.x - units, this.y);
            case Game_1.Game.DIRECTIONS.CENTER:
                return new Position(this.x, this.y);
        }
    };
    Position.prototype.equals = function (other) {
        return this.x == other.x && this.y == other.y;
    };
    Position.prototype.toString = function () {
        return this.x + "," + this.y;
    };
    return Position;
}());
exports.Position = Position;
