// @lux-ai/2021-challenge/lib/es5/Actions/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResearchAction = exports.PillageAction = exports.TransferAction = exports.SpawnCityAction = exports.SpawnWorkerAction = exports.SpawnCartAction = exports.SpawnAction = exports.MoveAction = exports.Action = void 0;
var Unit_1 = require("../Unit");
/**
 * Internal representations of agent sent actions. All actions must be validated prior to construction
 */
var Action = /** @class */ (function () {
    function Action(action, team) {
        this.action = action;
        this.team = team;
    }
    return Action;
}());
exports.Action = Action;
var MoveAction = /** @class */ (function (_super) {
    __extends(MoveAction, _super);
    function MoveAction(action, team, unitid, direction, newcell) {
        var _this = _super.call(this, action, team) || this;
        _this.unitid = unitid;
        _this.direction = direction;
        _this.newcell = newcell;
        return _this;
    }
    return MoveAction;
}(Action));
exports.MoveAction = MoveAction;
var SpawnAction = /** @class */ (function (_super) {
    __extends(SpawnAction, _super);
    function SpawnAction(action, team, x, y) {
        var _this = _super.call(this, action, team) || this;
        _this.x = x;
        _this.y = y;
        return _this;
    }
    return SpawnAction;
}(Action));
exports.SpawnAction = SpawnAction;
var SpawnCartAction = /** @class */ (function (_super) {
    __extends(SpawnCartAction, _super);
    function SpawnCartAction(action, team, x, y) {
        var _this = _super.call(this, action, team, x, y) || this;
        _this.type = Unit_1.Unit.Type.CART;
        return _this;
    }
    return SpawnCartAction;
}(SpawnAction));
exports.SpawnCartAction = SpawnCartAction;
var SpawnWorkerAction = /** @class */ (function (_super) {
    __extends(SpawnWorkerAction, _super);
    function SpawnWorkerAction(action, team, x, y) {
        var _this = _super.call(this, action, team, x, y) || this;
        _this.type = Unit_1.Unit.Type.WORKER;
        return _this;
    }
    return SpawnWorkerAction;
}(SpawnAction));
exports.SpawnWorkerAction = SpawnWorkerAction;
var SpawnCityAction = /** @class */ (function (_super) {
    __extends(SpawnCityAction, _super);
    function SpawnCityAction(action, team, unitid) {
        var _this = _super.call(this, action, team) || this;
        _this.unitid = unitid;
        return _this;
    }
    return SpawnCityAction;
}(Action));
exports.SpawnCityAction = SpawnCityAction;
var TransferAction = /** @class */ (function (_super) {
    __extends(TransferAction, _super);
    function TransferAction(action, team, srcID, destID, resourceType, amount) {
        var _this = _super.call(this, action, team) || this;
        _this.srcID = srcID;
        _this.destID = destID;
        _this.resourceType = resourceType;
        _this.amount = amount;
        return _this;
    }
    return TransferAction;
}(Action));
exports.TransferAction = TransferAction;
var PillageAction = /** @class */ (function (_super) {
    __extends(PillageAction, _super);
    function PillageAction(action, team, unitid) {
        var _this = _super.call(this, action, team) || this;
        _this.unitid = unitid;
        return _this;
    }
    return PillageAction;
}(Action));
exports.PillageAction = PillageAction;
var ResearchAction = /** @class */ (function (_super) {
    __extends(ResearchAction, _super);
    function ResearchAction(action, team, x, y) {
        var _this = _super.call(this, action, team) || this;
        _this.x = x;
        _this.y = y;
        return _this;
    }
    return ResearchAction;
}(Action));
exports.ResearchAction = ResearchAction;
