// dimensions-ai/lib/main/Tournament/RankSystem/TrueSkillSystem/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var sprintf_js_1 = require("sprintf-js");
var ts_trueskill_1 = require("ts-trueskill");
var DeepMerge_1 = require("../../../utils/DeepMerge");
var DeepCopy_1 = require("../../../utils/DeepCopy");
var __1 = require("..");
var TrueSkillSystem = /** @class */ (function (_super) {
    __extends(TrueSkillSystem, _super);
    function TrueSkillSystem(configs) {
        var _this = _super.call(this) || this;
        _this.configs = {
            initialMu: 25,
            initialSigma: 25 / 3,
        };
        _this.configs = DeepMerge_1.deepMerge(_this.configs, DeepCopy_1.deepCopy(configs));
        return _this;
    }
    TrueSkillSystem.prototype.initializeRankState = function () {
        return {
            rating: {
                mu: this.configs.initialMu,
                sigma: this.configs.initialSigma,
                score: this.getScore(this.configs.initialMu, this.configs.initialSigma),
            },
        };
    };
    TrueSkillSystem.prototype.getScore = function (mu, sigma) {
        return mu - 3 * sigma;
    };
    TrueSkillSystem.prototype.onPlayerUpdate = function (rankState) {
        return {
            rating: {
                mu: rankState.rating.mu,
                sigma: this.configs.initialSigma,
                score: this.getScore(rankState.rating.mu, this.configs.initialSigma),
            },
        };
    };
    TrueSkillSystem.prototype.updateRanks = function (rankStates, relativeRanks) {
        var _this = this;
        var ratings = [];
        rankStates.forEach(function (rankState) {
            ratings.push([new ts_trueskill_1.Rating(rankState.rating.mu, rankState.rating.sigma)]);
        });
        var newRatings = ts_trueskill_1.rate(ratings, relativeRanks);
        var newRankStates = [];
        newRatings.forEach(function (rating) {
            newRankStates.push({
                rating: {
                    mu: rating[0].mu,
                    sigma: rating[0].sigma,
                    score: _this.getScore(rating[0].mu, rating[0].sigma),
                },
            });
        });
        return newRankStates;
    };
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    TrueSkillSystem.prototype.resetRank = function (rankState) {
        return {
            rating: {
                mu: this.configs.initialMu,
                sigma: this.configs.initialSigma,
                score: this.getScore(this.configs.initialMu, this.configs.initialSigma),
            },
        };
    };
    TrueSkillSystem.prototype.rankComparator = function (rankState1, rankState2) {
        return rankState2.rating.score - rankState1.rating.score;
    };
    TrueSkillSystem.prototype.getRankStatesHeaderString = function () {
        return sprintf_js_1.sprintf("%-30s | %-14s | %-15s | %-18s | %-8s".underline, 'Name', 'ID', 'Score=(μ - 3σ)', 'Mu: μ, Sigma: σ', 'Matches');
    };
    TrueSkillSystem.prototype.getRankStateString = function (player, rankState, matchesPlayed) {
        return sprintf_js_1.sprintf("%-30s".blue +
            " | %-14s | " +
            "%-15s".green +
            " | " +
            "\u03BC=%-6s, \u03C3=%-6s".yellow +
            " | %-8s", player.tournamentID.name + (player.disabled ? ' X' : ''), player.tournamentID.id, (rankState.rating.mu - rankState.rating.sigma * 3).toFixed(7), rankState.rating.mu.toFixed(3), rankState.rating.sigma.toFixed(3), matchesPlayed);
    };
    return TrueSkillSystem;
}(__1.RankSystem));
exports.TrueSkillSystem = TrueSkillSystem;
