// dimensions-ai/lib/main/Tournament/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var events_1 = require("events");
var Match_1 = require("../Match");
var DimensionError_1 = require("../DimensionError");
var Logger_1 = require("../Logger");
var RankSystem_1 = require("./RankSystem");
var DeepCopy_1 = require("../utils/DeepCopy");
var utils_1 = require("../utils");
var System_1 = require("../utils/System");
var VarLock_1 = __importDefault(require("../utils/VarLock"));
var TournamentStatus_1 = require("./TournamentStatus");
var TournamentTypes_1 = require("./TournamentTypes");
// import TournamentStatusDefault = require('./TournamentStatus');
// import TournamentTypeDefault = require('./TournamentTypes');
/** @ignore */
// import _RankSystem = RankSystemDefault.RankSystem;
/** @ignore */
// import _TOURNAMENT_TYPE = TournamentTypeDefault.TournamentType;
/** @ignore */
// import _TournamentStatus = TournamentStatusDefault.TournamentStatus;
/**
 * Player class that persists data for the same ephemereal agent across multiple matches. Used for {@link Tournament | Tournaments}
 */
var Player = /** @class */ (function () {
    function Player(tournamentID, file, zipFile, botkey) {
        this.tournamentID = tournamentID;
        this.file = file;
        /**
         * Whether this player is anonymous and not tied to a user on the back end
         *
         * If this is ever false, then that means 1. we have a backend setup 2. there is an actual user entry
         */
        this.anonymous = true;
        /** Associated username if there is one */
        this.username = undefined;
        /**
         * Path to player's directory, not the file to be executed/used. Value is necessary if storage system is not used or
         * there is no bot key because the player is anonymous (locally stored). Used so tournament can clean up old bot
         * directory when a new bot is uploaded
         */
        this.botDirPath = undefined;
        /**
         * Key that references the player's bot file object if it exists
         */
        this.botkey = undefined;
        /**
         * Whether or not this player is disabled and won't be used in in the default match scheduling for
         * {@link Tournament.Ladder | Ladder Tournaments}. Is set to true if this player's bot throws an error during
         * the initialization stage of a {@link Match}.
         */
        this.disabled = false;
        /**
         * Path to the zip file for the bot. Used when no storage service is used. Used to allow api to send zipped bot file
         */
        this.zipFile = undefined;
        /**
         * The version of the bot associated with this player. Incremented whenever addPlayer is called
         */
        this.version = 0;
        this.botkey = botkey;
        this.zipFile = zipFile;
    }
    /**
     * Generates a 12 character player id string
     */
    Player.generatePlayerID = function () {
        return utils_1.genID(12);
    };
    return Player;
}());
exports.Player = Player;
/**
 * The tournament class and module extended by all concrete Tournament Classes. Tournament Types available now are
 * {@link Ladder}, and {@link Elimination}. A tournament is composed of players, which can either be
 * all locally stored, or a split between locally stored anonymous players and database stored user owned players.
 * Ladder is the only tournament where it can be made distributed, other tournament types may only be run as a single
 * instance
 *
 * Notes: `this.competitors` map is used when no DB is used. When a DB is used, locally stored players are only in
 * `this.anonymousCompetitors` and other players are pulled from DB. Hence, a lot of code requires checking if database
 * exists and if so, pull from there and the anonymous competitors map, other wise use this.state or this.competitors
 */
var Tournament = /** @class */ (function (_super) {
    __extends(Tournament, _super);
    function Tournament(design, id, tournamentConfigs, dimension) {
        var _this = _super.call(this) || this;
        _this.design = design;
        /** Mapping match ids to active ongoing matches */
        _this.matches = new Map();
        /** A queue whose elements are each arrays of players that are to compete against each other */
        _this.matchQueue = [];
        /** The current status of the tournament */
        _this.status = Tournament.Status.UNINITIALIZED;
        /** Logger */
        _this.log = new Logger_1.Logger();
        /** Registered competitors in this tournament */
        _this.competitors = new Map();
        /**
         * All competitors that are anonymous, local (no user), competitors and not registered in database. Used only when
         * there is a DB
         */
        _this.anonymousCompetitors = new Map();
        /**
         * This Tournament's name
         */
        _this.name = '';
        /**
         * Promise array of which all resolves once every player added through constructor is finished adding
         */
        _this.initialAddPlayerPromises = [];
        /**
         * Map from player ids to a promise that resolves once it is unlocked from the function that initiated a promise
         */
        _this.lockedPlayers = new Map();
        /**
         * Map from player ids to the last version used in a match in this tournament instance
         *
         * TODO: This is not a good solution, it is effectively an in-memory cache used to invalidate past bot files. It has
         * low space overhead but it won't scale nicely.
         */
        _this.lastVersionUsed = new Map();
        _this.id = id;
        // use overriden id if provided
        if (tournamentConfigs.id) {
            _this.id = tournamentConfigs.id;
        }
        _this.log.level =
            tournamentConfigs.loggingLevel !== undefined
                ? tournamentConfigs.loggingLevel
                : Logger_1.Logger.LEVEL.INFO;
        _this.name = tournamentConfigs.name
            ? tournamentConfigs.name
            : "tournament_" + _this.id;
        _this.log.identifier = _this.name;
        _this.dimension = dimension;
        _this.log.info("Created Tournament - ID: " + _this.id + ", Name: " + _this.name);
        // if no name is provided but database is being used, log an error
        if (!tournamentConfigs.name && dimension.hasDatabase()) {
            _this.log.error("A name has to be specified for a tournament otherwise tournament player data will not be reused across runs of the tournament");
        }
        return _this;
    }
    /**
     * Add or update a player to the tournament
     *
     * If no existing ID is specified, this is treated as adding a completely new player.
     *
     * If existing ID is specified and that ID exists already, this will update the file for that player instead and
     * effectively update the player. First time a player is added, if there is existing stats in a DB they won't be
     * reset. Subsequent adds will change the stats.
     *
     * If the player is to exist beyond the tournament, an existingID must always be provided and generated somewhere else
     *
     * Resolves with the new player or updated player
     *
     * @param file - The file to the bot or an object with the file and a name for the player specified
     * @param existingID - The optional id of the player. Can also be provided in the first arg in a object
     * @param calledFromInitialization - Whether or not the player was called from initialization. If true, player
     * version does not increment
     *
     */
    Tournament.prototype.addplayer = function (file, existingID, calledFromInitialization) {
        if (calledFromInitialization === void 0) { calledFromInitialization = false; }
        return __awaiter(this, void 0, void 0, function () {
            var id, playerStat, varlock, player, oldname, oldfile, err_1, name_1, newPlayer, user, newPlayer, user;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (typeof file !== 'string') {
                            if (!existingID) {
                                existingID = file.existingID;
                            }
                        }
                        if (!existingID) return [3 /*break*/, 11];
                        return [4 /*yield*/, this.getPlayerStat(existingID)];
                    case 1:
                        playerStat = (_a.sent()).playerStat;
                        if (!playerStat) return [3 /*break*/, 9];
                        varlock = this.lockedPlayers.get(existingID);
                        if (!varlock) return [3 /*break*/, 3];
                        return [4 /*yield*/, varlock.lockvar()];
                    case 2:
                        _a.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        varlock = new VarLock_1.default();
                        this.lockedPlayers.set(existingID, varlock);
                        _a.label = 4;
                    case 4:
                        player = playerStat.player;
                        // undisable the player
                        player.disabled = false;
                        oldname = player.tournamentID.name;
                        oldfile = player.file;
                        // remove the oldfile
                        if (player.botDirPath) {
                            System_1.removeDirectorySync(player.botDirPath);
                        }
                        if (typeof file === 'string') {
                            player.file = file;
                        }
                        else {
                            player.file = file.file;
                            player.tournamentID.name = file.name;
                            player.botDirPath = file.botdir;
                            player.zipFile = file.zipFile;
                            player.botkey = file.botkey;
                            if (!calledFromInitialization) {
                                // update version and last update time stamp
                                player.version++;
                            }
                        }
                        _a.label = 5;
                    case 5:
                        _a.trys.push([5, 7, , 8]);
                        return [4 /*yield*/, this.updatePlayer(player, oldname, oldfile)];
                    case 6:
                        _a.sent();
                        varlock.unlock();
                        this.lockedPlayers.delete(existingID);
                        return [3 /*break*/, 8];
                    case 7:
                        err_1 = _a.sent();
                        varlock.unlockWithError(err_1);
                        this.lockedPlayers.delete(existingID);
                        throw err_1;
                    case 8:
                        id = existingID;
                        return [2 /*return*/, player];
                    case 9:
                        // otherwise bot doesn't exist locally or in db row statistics field, and we use this id as our id to generate a new player
                        id = existingID;
                        _a.label = 10;
                    case 10: return [3 /*break*/, 12];
                    case 11:
                        id = this.generateNextTournamentIDString();
                        _a.label = 12;
                    case 12:
                        if (!(typeof file === 'string')) return [3 /*break*/, 17];
                        name_1 = "player-" + id;
                        newPlayer = new Player({ id: id, name: name_1, username: undefined }, file, undefined);
                        if (!this.dimension.hasDatabase()) return [3 /*break*/, 14];
                        return [4 /*yield*/, this.dimension.databasePlugin.getUser(newPlayer.tournamentID.id)];
                    case 13:
                        user = _a.sent();
                        if (user) {
                            newPlayer.anonymous = false;
                            newPlayer.username = user.username;
                            newPlayer.tournamentID.username = user.username;
                        }
                        else {
                            this.competitors.set(id, newPlayer);
                        }
                        return [3 /*break*/, 15];
                    case 14:
                        this.competitors.set(id, newPlayer);
                        _a.label = 15;
                    case 15:
                        if (newPlayer.anonymous) {
                            this.anonymousCompetitors.set(id, newPlayer);
                        }
                        return [4 /*yield*/, this.internalAddPlayer(newPlayer)];
                    case 16:
                        _a.sent();
                        return [2 /*return*/, newPlayer];
                    case 17:
                        newPlayer = new Player({ id: id, name: file.name, username: undefined }, file.file, file.zipFile, file.botkey);
                        newPlayer.botDirPath = file.botdir;
                        if (!this.dimension.hasDatabase()) return [3 /*break*/, 19];
                        return [4 /*yield*/, this.dimension.databasePlugin.getUser(newPlayer.tournamentID.id)];
                    case 18:
                        user = _a.sent();
                        if (user) {
                            newPlayer.tournamentID.name = file.name;
                            newPlayer.anonymous = false;
                            newPlayer.username = user.username;
                            newPlayer.tournamentID.username = user.username;
                        }
                        else {
                            this.competitors.set(id, newPlayer);
                        }
                        return [3 /*break*/, 20];
                    case 19:
                        this.competitors.set(id, newPlayer);
                        _a.label = 20;
                    case 20:
                        if (newPlayer.anonymous) {
                            this.anonymousCompetitors.set(id, newPlayer);
                        }
                        return [4 /*yield*/, this.internalAddPlayer(newPlayer)];
                    case 21:
                        _a.sent();
                        return [2 /*return*/, newPlayer];
                }
            });
        });
    };
    /**
     * Returns a new id for identifying a player in a tournament
     * Only used when adding a plyaer to a tournament is done without specifying an id to use.
     */
    Tournament.prototype.generateNextTournamentIDString = function () {
        return Player.generatePlayerID();
    };
    /**
     * Disables the player with id playerID
     * @param playerID - the player's id to disable
     */
    Tournament.prototype.disablePlayer = function (playerID) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, user, playerStat;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.getPlayerStat(playerID)];
                    case 1:
                        _a = _b.sent(), user = _a.user, playerStat = _a.playerStat;
                        if (!playerStat) return [3 /*break*/, 4];
                        playerStat.player.disabled = true;
                        if (!(this.dimension.hasDatabase() && user)) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.dimension.databasePlugin.updateUser(playerID, user)];
                    case 2:
                        _b.sent();
                        _b.label = 3;
                    case 3: return [3 /*break*/, 5];
                    case 4: throw new DimensionError_1.TournamentPlayerDoesNotExistError("Player " + playerID + " was not found in this tournament");
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Removes the competitor/player with id `playerID` (a {@link nanoid}). Resolves if succesful, otherwise rejects if
     * player doesn't exist or couldn't be removed
     *
     * @param playerID - ID of the player to remove
     */
    Tournament.prototype.removePlayer = function (playerID) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, user, playerStat;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.getPlayerStat(playerID)];
                    case 1:
                        _a = _b.sent(), user = _a.user, playerStat = _a.playerStat;
                        if (!playerStat) return [3 /*break*/, 5];
                        this.competitors.delete(playerID);
                        this.anonymousCompetitors.delete(playerID);
                        // disable player
                        playerStat.player.disabled = true;
                        if (!(this.dimension.hasDatabase() && user)) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.dimension.databasePlugin.updateUser(playerID, user)];
                    case 2:
                        _b.sent();
                        _b.label = 3;
                    case 3: return [4 /*yield*/, this.internalRemovePlayer(playerID)];
                    case 4:
                        _b.sent();
                        return [3 /*break*/, 6];
                    case 5: throw new DimensionError_1.TournamentPlayerDoesNotExistError('Not a player');
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
    Tournament.prototype.internalRemovePlayer = function (playerID) {
        return __awaiter(this, void 0, void 0, function () { return __generator(this, function (_a) {
            return [2 /*return*/];
        }); });
    };
    /**
     * Runs a match
     * @param players - the players to compete together
     * @returns a promise that resolves with the results and the associated match
     */
    Tournament.prototype.runMatch = function (players) {
        return __awaiter(this, void 0, void 0, function () {
            var locks, matchConfigs, agentSpecificOptions, _i, players_1, player, lv, filesAndNamesAndIDs, match, i, varlock, results, err_2;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!players.length)
                            throw new DimensionError_1.FatalError('No players provided for match');
                        locks = [];
                        players.forEach(function (player) {
                            if (_this.lockedPlayers.has(player.tournamentID.id)) {
                                locks.push(_this.lockedPlayers.get(player.tournamentID.id).lock);
                            }
                        });
                        matchConfigs = DeepCopy_1.deepCopy(this.getConfigs().defaultMatchConfigs);
                        agentSpecificOptions = [];
                        _i = 0, players_1 = players;
                        _a.label = 1;
                    case 1:
                        if (!(_i < players_1.length)) return [3 /*break*/, 4];
                        player = players_1[_i];
                        return [4 /*yield*/, this.getLastVersionUsedOfPlayer(player.tournamentID.id)];
                    case 2:
                        lv = _a.sent();
                        if (lv !== undefined && lv === player.version) {
                            this.log.system("player " + player.tournamentID.id + " attempting to use cached bot");
                            agentSpecificOptions.push({
                                useCachedBotFile: true,
                            });
                        }
                        else {
                            this.log.system("player " + player.tournamentID.id + " not using cached bot file");
                            agentSpecificOptions.push({
                                useCachedBotFile: false,
                            });
                        }
                        _a.label = 3;
                    case 3:
                        _i++;
                        return [3 /*break*/, 1];
                    case 4:
                        matchConfigs.agentSpecificOptions = agentSpecificOptions;
                        filesAndNamesAndIDs = players.map(function (player) {
                            // if player has a botkey, use that, otherwise use whats in player.file
                            return {
                                file: player.file,
                                tournamentID: player.tournamentID,
                                botkey: player.botkey,
                                version: player.version,
                            };
                        });
                        match = new Match_1.Match(this.design, filesAndNamesAndIDs, matchConfigs, this.dimension);
                        // store match into the tournament locally
                        this.matches.set(match.id, match);
                        _a.label = 5;
                    case 5:
                        _a.trys.push([5, 14, 15, 16]);
                        /**
                         * using "mutex" locks to force any update to players to happen completely before or after match initialization,
                         * helping eliminate race conditions
                         */
                        // wait for all locks to release
                        return [4 /*yield*/, Promise.all(locks)];
                    case 6:
                        /**
                         * using "mutex" locks to force any update to players to happen completely before or after match initialization,
                         * helping eliminate race conditions
                         */
                        // wait for all locks to release
                        _a.sent();
                        i = 0;
                        _a.label = 7;
                    case 7:
                        if (!(i < players.length)) return [3 /*break*/, 11];
                        varlock = this.lockedPlayers.get(players[i].tournamentID.id);
                        if (!varlock) return [3 /*break*/, 9];
                        return [4 /*yield*/, varlock.lockvar()];
                    case 8:
                        _a.sent();
                        return [3 /*break*/, 10];
                    case 9:
                        this.lockedPlayers.set(players[i].tournamentID.id, new VarLock_1.default());
                        _a.label = 10;
                    case 10:
                        i++;
                        return [3 /*break*/, 7];
                    case 11: return [4 /*yield*/, match.initialize().finally(function () {
                            // release all locks once this is done
                            for (var i = 0; i < players.length; i++) {
                                var varlock = _this.lockedPlayers.get(players[i].tournamentID.id);
                                if (varlock) {
                                    varlock.unlock();
                                    // delete the varlock
                                    _this.lockedPlayers.delete(players[i].tournamentID.id);
                                }
                            }
                        })];
                    case 12:
                        _a.sent();
                        // note: method to test locks, match.initialize(); do a await sleep(5000); and try updating bot during initialization. If response takes a few seconds to respond, then locks worked as update happenedd after initialization.
                        players.forEach(function (player) { return __awaiter(_this, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, this.updateLastVersionUsedOfPlayer(player.tournamentID.id, player.version)];
                                    case 1:
                                        _a.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); });
                        return [4 /*yield*/, match.run()];
                    case 13:
                        results = _a.sent();
                        // remove the match from the active matches list
                        this.matches.delete(match.id);
                        // Resolve the results
                        return [2 /*return*/, { results: results, match: match }];
                    case 14:
                        err_2 = _a.sent();
                        this.emit(Tournament.Events.MATCH_RAN);
                        return [2 /*return*/, {
                                results: false,
                                err: err_2,
                                match: match,
                            }];
                    case 15:
                        // regardless of error or not ensure match is stored
                        // if database plugin is active and saveTournamentMatches is set to true, store match
                        if (this.dimension.hasDatabase()) {
                            if (this.dimension.databasePlugin.configs.saveTournamentMatches) {
                                this.dimension.databasePlugin.storeMatch(match, this.id);
                            }
                        }
                        return [7 /*endfinally*/];
                    case 16: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Return an Array of Players corresponding to the player ids stored in `queuedMatchInfo`
     * @param queuedMatchInfo
     */
    Tournament.prototype.getMatchInfoFromQueuedMatch = function (queuedMatchInfo) {
        return __awaiter(this, void 0, void 0, function () {
            var retrievePlayerPromises, i, playerId;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        retrievePlayerPromises = [];
                        for (i = 0; i < queuedMatchInfo.length; i++) {
                            playerId = queuedMatchInfo[i];
                            retrievePlayerPromises.push(this.getPlayerStat(playerId).then(function (_a) {
                                var playerStat = _a.playerStat;
                                return playerStat.player;
                            }));
                        }
                        return [4 /*yield*/, Promise.all(retrievePlayerPromises)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Tournament.prototype.getLastVersionUsedOfPlayer = function (playerID) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, Promise.resolve(this.lastVersionUsed.get(playerID))];
            });
        });
    };
    Tournament.prototype.updateLastVersionUsedOfPlayer = function (playerID, version) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.lastVersionUsed.set(playerID, version);
                return [2 /*return*/];
            });
        });
    };
    /**
     * Removes a match by id. Returns true if deleted, false if nothing was deleted
     */
    Tournament.prototype.removeMatch = function (matchID) {
        return __awaiter(this, void 0, void 0, function () {
            var match;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.matches.has(matchID)) return [3 /*break*/, 2];
                        match = this.matches.get(matchID);
                        return [4 /*yield*/, match.destroy()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, this.matches.delete(matchID)];
                    case 2: return [2 /*return*/, false];
                }
            });
        });
    };
    /**
     * Destroy this tournament. Rejects if an error occured in trying to destroy it.
     */
    Tournament.prototype.destroy = function () {
        return __awaiter(this, void 0, void 0, function () {
            var destroyPromises;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.preInternalDestroy()];
                    case 1:
                        _a.sent();
                        // stop if running
                        if (this.status === Tournament.Status.RUNNING)
                            this.stop();
                        destroyPromises = [];
                        // now remove all match processes
                        this.matches.forEach(function (match) {
                            destroyPromises.push(match.destroy());
                        });
                        return [4 /*yield*/, Promise.all(destroyPromises)];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, this.postInternalDestroy()];
                    case 3:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Pre run function before generic destroy takes place
     */
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    Tournament.prototype.preInternalDestroy = function () {
        return __awaiter(this, void 0, void 0, function () { return __generator(this, function (_a) {
            return [2 /*return*/];
        }); });
    };
    /**
     * Post run function before generic destroy takes place
     */
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    Tournament.prototype.postInternalDestroy = function () {
        return __awaiter(this, void 0, void 0, function () { return __generator(this, function (_a) {
            return [2 /*return*/];
        }); });
    };
    /**
     * Generates a 6 character tournament ID identifying this tournament class instance. Not to be confused with
     * {@link Tournament.ID} which is the ID for competitors in the tournament
     */
    Tournament.genTournamentClassID = function () {
        return utils_1.genID(6);
    };
    /**
     * Returns the name of the tournament but formatted (no spaces)
     */
    Tournament.prototype.getSafeName = function () {
        return this.name.replace(/ /g, '_');
    };
    /**
     * Returns a key name to be used when storing a tournament by a combination of its name and id
     */
    Tournament.prototype.getKeyName = function () {
        return this.getSafeName() + "_" + this.id;
    };
    /**
     * Resolves with player stats if player with the id exists. Includes database user if db contains the player
     * Fields are null if they don't exist. If playerStat field is null, then this player does not exist
     *
     * @param id - id of player to get
     */
    Tournament.prototype.getPlayerStat = function (id) {
        return __awaiter(this, void 0, void 0, function () {
            var user, err_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!this.state.playerStats.has(id)) return [3 /*break*/, 6];
                        if (!this.dimension.hasDatabase()) return [3 /*break*/, 5];
                        user = void 0;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.dimension.databasePlugin.getUser(id)];
                    case 2:
                        user = _a.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        err_3 = _a.sent();
                        this.log.error(err_3);
                        throw new DimensionError_1.DatabaseGetUserError("failed to get database user with id: " + id);
                    case 4:
                        if (user && user.statistics[this.getKeyName()]) {
                            return [2 /*return*/, { user: user, playerStat: user.statistics[this.getKeyName()] }];
                        }
                        _a.label = 5;
                    case 5: return [3 /*break*/, 7];
                    case 6: return [2 /*return*/, { user: null, playerStat: this.state.playerStats.get(id) }];
                    case 7: return [2 /*return*/, { user: null, playerStat: null }];
                }
            });
        });
    };
    return Tournament;
}(events_1.EventEmitter));
exports.Tournament = Tournament;
// some imports moved to here to avoid circular issues with using values
var Ladder_1 = require("./Ladder");
/** @ignore */
// import LadderTournament = LadderDefault.Ladder;
var Elimination_1 = require("./Elimination");
/** @ignore */
// import EliminationTournament = EliminationDefault.Elimination;
var Scheduler_1 = require("./Scheduler");
/** @ignore */
// import SchedulerClass = SchedulerDefault.Scheduler;
(function (Tournament) {
    // Re-export tournament classes/namespaces
    /* eslint-disable */
    Tournament.Ladder = Ladder_1.Ladder;
    Tournament.Elimination = Elimination_1.Elimination;
    Tournament.Scheduler = Scheduler_1.Scheduler;
    // Re-export some types
    Tournament.Type = TournamentTypes_1.TournamentType;
    Tournament.Status = TournamentStatus_1.TournamentStatus;
    Tournament.RankSystem = RankSystem_1.RankSystem;
    var Events;
    (function (Events) {
        /**
         * Event is emitted when initialAddPlayerPromises resolves. This involves all players initialized to tournament
         * upon initialization of tournament instance
         */
        Events["INITIAL_PLAYERS_INITIALIZED"] = "initial_players_initialized";
        /**
         * Event is emitted whenever a match runs
         */
        Events["MATCH_RAN"] = "match_ran";
        /**
         * Event is emitted whenever a match runs AND the tournament is completely done handling it.
         */
        Events["MATCH_HANDLED"] = "match_handled";
    })(Events = Tournament.Events || (Tournament.Events = {}));
    var RankSystemTypes;
    (function (RankSystemTypes) {
        /** Ranking by wins, ties and losses */
        RankSystemTypes["WINS"] = "wins";
        /** Ranking by the ELO ranking system */
        RankSystemTypes["ELO"] = "elo";
        /** Ranking by Microsoft's Trueskill */
        RankSystemTypes["TRUESKILL"] = "trueskill";
    })(RankSystemTypes = Tournament.RankSystemTypes || (Tournament.RankSystemTypes = {}));
})(Tournament = exports.Tournament || (exports.Tournament = {}));
exports.Tournament = Tournament;
